src/emulator/storage/files.ts

```diff
@@ -291,11 +300,22 @@ export class StorageLayer {
     const file = new StoredFile(finalMetadata, filePath);
     this._files.set(filePath, file);
 
-    this._persistence.deleteFile(filePath, true);
+    return { upload, file };
+  }
+
+  /**
+   * Updates the uploaded file's location in persistence layer and triggers CREATE Cloud Functions.
+   * To be called as the last step of finalizing an upload.
+   * @param finishedUpload The upload to finalize and its corresponding file.
+   */
+  public persistsUpload(finishedUpload: FinishedUpload): void {
```

Comment:
IMO this still violates the easy to use, hard to misuse principle, particularly the "hard to misuse" part.

If a caller calls `storeFinishedUpload`  and forgets to call `persistsUpload`, we end up in a state where the file is registered in `this._files` but not saved at the final filepath (`this.path(upload.bucketId, upload.objectId)`). Trying to export in this state cause errors. If auth fails, we will also be in a state where the file gets deleted but the ResumableUpload has been marked as successful.

It seems we split `finalizeUpload` into two methods so that we can continue to use `StorageLayer.getMetadata` in the finalize upload handler, but this confuses the semantics of what a `StoredFile` is a bit.  

Can you think of another way to get file metadata for rules eval without persisting the file? What if we decouple the metadata logic from `this._files`? what if we did something like `isPermitted(Metadata.getMetadataForUpload(uploadId), ...)`

Reply:
+1 I think this is blocking

Reply:
Ok, I'll make a new method `createMetadata` on `StorageLayer`.